presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3ClientFactory.java

```diff
@@ -0,0 +1,184 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.s3;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.Protocol;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.metrics.RequestMetricCollector;
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Builder;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.facebook.presto.hive.HiveClientConfig;
+import io.airlift.units.Duration;
+import org.apache.hadoop.conf.Configuration;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration;
+import static com.amazonaws.regions.Regions.US_EAST_1;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_ENDPOINT;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_PIN_CLIENT_TO_CURRENT_REGION;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+
+/**
+ * This factory provides AmazonS3 client required for executing S3SelectPushdown requests.
+ * Normal S3 GET requests use AmazonS3 clients initialized in PrestoS3FileSystem or EMRFS.
+ * The ideal state will be to merge this logic with the two file systems and get rid of this
+ * factory class.
+ * Please do not use the client provided by this factory for any other use cases.
+ */
+public class PrestoS3ClientFactory
+{
+    private static final String S3_ACCESS_KEY = "presto.s3.access-key";
+    private static final String S3_SECRET_KEY = "presto.s3.secret-key";
+    private static final String S3_CREDENTIALS_PROVIDER = "presto.s3.credentials-provider";
+    private static final String S3_USE_INSTANCE_CREDENTIALS = "presto.s3.use-instance-credentials";
+    private static final String S3_CONNECT_TIMEOUT = "presto.s3.connect-timeout";
+    private static final String S3_SOCKET_TIMEOUT = "presto.s3.socket-timeout";
+    private static final String S3_SSL_ENABLED = "presto.s3.ssl.enabled";
+    private static final String S3_MAX_ERROR_RETRIES = "presto.s3.max-error-retries";
+    private static final String S3_USER_AGENT_PREFIX = "presto.s3.user-agent-prefix";
+    private static final String S3_SELECT_PUSHDOWN_MAX_CONNECTIONS = "hive.s3select-pushdown.max-connections";
+    private static String s3UserAgentSuffix = "presto";
+    private volatile AmazonS3 s3Client;
+
+    AmazonS3 getS3Client(Configuration config, HiveClientConfig clientConfig)
```

Comment:
I found the synchronization in this class confusing. If we also make this method synchronized we don't need to mark `s3Client` volatile (if you are worried about performance of synchronizing this, I don't think we will see any perf. difference),  and we can get rid of the `if` check at the beginning of `initializeS3Client`. Also `initializeS3Client` can return void as it just sets the `se3Client` field.

After that change we should also mark `s3Client` as:

```
@GuardedBy("this")
private AmazonS3 s3Client;
```

So, overall I propose:

```
@GuardedBy("this")
private AmazonS3 s3Client;

synchronized AmazonS3 getS3Client(Configuration config, HiveClientConfig clientConfig)
{
    if (s3Client == null) {
        initializeS3Client(config, clientConfig);
    }
    return s3Client;
}

synchronized void initializeS3Client(Configuration config, HiveClientConfig clientConfig)
{
```

Reply:
I am not a big fan of optimizing code *before* we prove/have data that shows clearly that there is a performance problem. That's why I proposed making this code simpler first. Later on if we have data that shows this synch. is a problem we can think of ways to optimize it.

Reply:
Makes sense.